lib/src/file_util.rs

```diff
@@ -134,6 +146,39 @@ pub fn persist_content_addressed_temp_file<P: AsRef<Path>>(
     }
 }
 
+/// Symlinks are always available on UNIX, but on Windows they require admin or
+/// Developer Mode
+#[cfg(windows)]
+pub fn try_symlink<P: AsRef<Path>, Q: AsRef<Path>>(
+    original: P,
+    link: Q,
+) -> Result<(), SymlinkError> {
+    const ERROR_PRIVILEGE_NOT_HELD: i32 = 1314;
+
+    if Path::is_dir(original.as_ref()) {
+        symlink_dir(original, link)
+    } else {
+        symlink_file(original, link)
+    }
```

Comment:
Perhaps, the `original` file kind should be passed by caller (based on the repository data, not on the actual filesystem state.) It's scary to access the `original` path here. The path may point to arbitrary filesystem location.

Oh, does it also mean the `original` path can't be switched between file and directory once the symlink is created?

Reply:
The way it works on Windows is that the symlink itself is to a directory or to a file. If you change the target to be the other thing, it just stops working until you change it back. Here's an example:
![3k9uqVx 1](https://github.com/martinvonz/jj/assets/1612285/a2ed86ee-ea84-49f9-bd41-dafbd5fe6ca6)

What kind of path-switching do you have in mind?

Reply:
> What kind of path-switching do you have in mind?

Directory->file on `jj new REV` for example. `jj` wouldn't recreate the symlink itself so long as the symlink data (the `original` path) is unchanged. If git were modeled after Windows symlinks, directory-symlink and file-symlink would be recorded differently, but it's not.

Reply:
I see the problem. I'll look at the tree data structures and see what it would take to pass the information through.

Reply:
I looked into this and, as you say, git doesn't store the information we'd need. It just has a relative path, and so that's all `MaterializedTreeValue::Symlink` contains; `TreeState::write_symlink` just receives an absolute and a relative path. We can still implement this, but it's a bit janky: 

`TreeState` knows it's writing a symlink of unknown valence, so it can determine whether it's intended to refer to a directory using `std::path::Path`, and pass this information to the low-level `try_symlink`. Unix can just ignore the flag. Windows has to test whether a symlink of the opposite type already exists, and if so, remove it; then, continue as before.

I've tested this approach and it works in practice. In some future where the backend has sufficiently rich information, we could extract the information from there instead of heuristically determining it in TreeState.

Reply:
> That's kind of true, but the use of Path::is_dir() has problems...

This feels like a Windows problem.  I wonder if it's feasible for jj to just create a "file" type symlink whenever it's unsure. We could do that if the link points outside the repo, as well.

For example, does something terrible happen on Windows if you create a symlink to a file and then replace the target file with a dir? Even if it does not work any longer, we can argue it's a Windows limitation we can't handle.

Reply:
> This feels like a Windows problem. I wonder if it's feasible for jj to just create a "file" type symlink whenever it's unsure. We could do that if the link points outside the repo, as well.

Yeah, I think that's pretty okay. I also think using normal file is good as the minimal support. It will avoid path separator `/` vs `\\` issues and weird symlinks like `///`.

Reply:
> For example, does something terrible happen on Windows if you create a symlink to a file and then replace the target file with a dir? Even if it does not work any longer, we can argue it's a Windows limitation we can't handle.

The symlink stops working, and JJ is unable to materialize or even `op restore`. You can recover from it, but only by deleting the bad symlink externally.